### PR TITLE
fix: replace unprintable and invalid characters in errors

### PR DESCRIPTION
**Pull request by davidby-influx**
Replace unprintable and invalid characters with '?'
in logged errors.  Truncate consecutive runs of them to
only 3 repeats of '?'

closes https://github.com/influxdata/influxdb/issues/23386

- [X] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0/)
- [X] Rebased/mergeable
- [X] Tests pass

### CODE REVIEW

File: tsdb/shard.go [hunk 1]
```diff
@@ -723,6 +724,39 @@ func (s *Shard) validateSeriesAndFields(points []models.Point, tracker StatsTrac
...
+const unPrintReplRune = '?'
+const unPrintMaxReplRune = 3
+
+// makePrintable - replace invalid and non-printable unicode characters with a few '?' runes
+func makePrintable(s string) string {
+	b := strings.Builder{}
+	b.Grow(len(s))
+	for _, r := range strings.ToValidUTF8(s, string(unPrintReplRune)) {
+		if !unicode.IsPrint(r) || r == unicode.ReplacementChar {
+			b.WriteRune(unPrintReplRune)
```
Comment (lesam) [inline tsdb/shard.go:736]: If you kept the 3-byte counter here instead, you wouldn't compress valid unicode `WHAT????` to `WHAT???`

Comment (davidby-influx) [reply tsdb/shard.go:736]: I also want to elide runs of '?' generated by `string.ToValidUTF8`, which in the customer scenario are in the hundreds of characters range (they put thousand-character runs of invalid Unicode characters in their line protocol).  The only complete solution for perfect operation I saw was to re-implement `strings.ToValidUTF8` with run-length detection, and that seemed overkill.

Comment (davidby-influx) [reply tsdb/shard.go]: Now fixed with `unicode.ReplacementChar` and a single replacement elision loop.

File: tsdb/shard.go [hunk 2]
```diff
@@ -731,29 +731,17 @@ const unPrintMaxReplRune = 3
...
+	c := 0
+	for _, r := range strings.ToValidUTF8(s, string(unicode.ReplacementChar)) {
 		if !unicode.IsPrint(r) || r == unicode.ReplacementChar {
-			b.WriteRune(unPrintReplRune)
+			if c < unPrintMaxReplRune {
+				b.WriteRune(unPrintReplRune)
+			}
+			c++
 		} else {
 			b.WriteRune(r)
```
Comment (lesam) [inline tsdb/shard.go:742]: `c=0` to reset here, so each group is limited to 3 instead of 3 total?

Comment (lesam) [reply tsdb/shard.go:742]: Should have a test for that case too I guess.

File: tsdb/shard_internal_test.go [hunk 3]
```diff
@@ -16,8 +16,32 @@ import (
...
+		string([]byte{'b', 'e', 'n', 't', 'e', 's', 't', '\t', '\n'}),
+		string([]byte{'b', 'e', 'n', 't', 'e', 's', 0, 0, 0xFE, 0, 0xFE, 't'}),
+		string([]byte{'b', 'e', 'n', 't', 'e', 's', 't', 0, 0, 0, 0, 0xFE, '\t', '\n', '\t', '\t', '\t'}),
+	}
+
+	for _, s := range badStrings {
+		f := makePrintable(s)
+		require.True(t, models.ValidKeyToken(f))
+		c := 0
+		for _, r := range f {
```
[the remark this answers is not in the document]
Comment (davidby-influx) [reply tsdb/shard_internal_test.go:38]: I wrote a sequence counter, but I can change it if you prefer.  Take a look at your convenience, no rush.

**Pull request merged**